File: 3.x.x/crates-utopia/emails/src/canonicals/icloud.rs
use crate::canonicals::provider::{
    remove_plus_addressing, require_non_empty_local, to_lower_case, Canonical, Provider,
};
use crate::EmailError;

/// iCloud / me.com / mac.com canonicalization (PHP `Icloud`).
#[derive(Debug, Clone, Copy, Default)]
pub struct Icloud;

const SUPPORTED_DOMAINS: &[&str] = &["icloud.com", "me.com", "mac.com"];
const CANONICAL_DOMAIN: &str = "icloud.com";

impl Provider for Icloud {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, _domain: &str) -> Result<Canonical, EmailError> {
        let normalized = to_lower_case(local);
        let normalized = remove_plus_addressing(&normalized);
        let normalized = require_non_empty_local(normalized)?;
        Ok(Canonical {
            local: normalized,
            domain: CANONICAL_DOMAIN.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Canonicals/Providers/Icloud.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * iCloud
 *
 * Handles Apple iCloud email normalization based on validator.js rules
 * - Removes plus addressing (subaddress)
 * - Preserves dots in local part
 * - Normalizes to icloud.com domain
 */
class Icloud extends Provider
{
    private const SUPPORTED_DOMAINS = ['icloud.com', 'me.com', 'mac.com'];

    private const CANONICAL_DOMAIN = 'icloud.com';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        // Remove plus addressing (subaddress) - everything after +
        $normalizedLocal = $this->removePlusAddressing($normalizedLocal);

        // Ensure local part is not empty after normalization
        if (empty($normalizedLocal)) {
            throw new \InvalidArgumentException('Email local part cannot be empty after normalization');
        }

        return [
            'local' => $normalizedLocal,
            'domain' => self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
